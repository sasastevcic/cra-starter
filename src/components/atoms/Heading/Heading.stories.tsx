import { ReactElement } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { FONT_SIZE_REGEX } from '../../../constants/regex';
import { typeStyle } from '../../../styles/config/typeStyles';
import { Table as GenericTable } from '../../../styles/storybook/Storybook';
import { StyledRow, StyledData } from '../../../styles/storybook/Storybook.styles';
import Paragraph from '../Paragraph';
import { HeadingProps, HeadingType } from './Heading.data';
import HeadingComponent from './index';

const Table = (props: HeadingProps): ReactElement => {
	const tableHeadings = ['Type', 'Size', 'Preview'];

	const rows = Object.entries(HeadingType).map(([name, type]) => {
		const fontSizesRegex = typeStyle[type].join('').match(FONT_SIZE_REGEX);

		const fontSizes = fontSizesRegex?.filter((fontSize) => !fontSize.includes('calc')) ?? [];

		if (fontSizes.length > 1) {
			tableHeadings[1] = 'Size (Fluid)';
		}

		return (
			<StyledRow key={name}>
				<StyledData>
					<Paragraph>{name}</Paragraph>
				</StyledData>
				<StyledData>
					<Paragraph>{fontSizes.join(' - ')}</Paragraph>
				</StyledData>
				<StyledData>
					<HeadingComponent type={type} {...props}>
						Heading
					</HeadingComponent>
				</StyledData>
			</StyledRow>
		);
	});

	return <GenericTable headings={tableHeadings}>{rows}</GenericTable>;
};

export default {
	title: 'Types/Headings',
	component: HeadingComponent,
} as ComponentMeta<typeof HeadingComponent>;

const Template: ComponentStory<typeof Table> = (args) => <Table {...args} />;

export const Headings = Template.bind({});
