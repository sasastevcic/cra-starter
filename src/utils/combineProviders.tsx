/* eslint-disable @typescript-eslint/no-explicit-any */
import { ComponentProps, ComponentType, FunctionComponent, ReactNode } from 'react';

type Providers = Array<[ComponentType<any>, ComponentProps<any>?]>;

export const combineProviders = (
	providers: Providers,
): FunctionComponent<{ children: ReactNode }> =>
	providers.reduce(
		(AccumulatedProviders, [Provider, props = {}]) =>
			({ children }) =>
				(
					<AccumulatedProviders>
						<Provider {...props}>{children}</Provider>
					</AccumulatedProviders>
				),
		({ children }) => <>{children}</>,
	);
