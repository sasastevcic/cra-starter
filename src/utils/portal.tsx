import { ReactNode, ReactPortal, useMemo } from 'react';
import { createPortal } from 'react-dom';

interface Portal {
	portalId: string;
	children: ReactNode;
}

const Portal = ({ portalId, children }: Portal): ReactPortal => {
	const portalElement = useMemo(() => document.getElementById(portalId), [portalId]);

	if (!portalElement) {
		throw new Error(`Element not found! Element id: ${portalId}`);
	}

	return createPortal(children, portalElement);
};

export default Portal;
