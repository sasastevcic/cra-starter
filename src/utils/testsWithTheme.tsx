/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ReactElement, ComponentType, ReactNode } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	// @ts-ignore
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType<{ children: ReactNode }> = ({ children }) =>
		componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
