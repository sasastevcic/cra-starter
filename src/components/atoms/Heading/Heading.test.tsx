import renderer from 'react-test-renderer';
import { renderWithTheme } from '../../../utils/testsWithTheme';
import Noop from '../Noop';
import { HeadingType } from './Heading.data';
import Heading from './index';

describe('Heading', () => {
	describe('Render', () => {
		it('should render', () => {
			const { getByTestId } = renderWithTheme(
				<Heading>
					<Noop />
				</Heading>,
			);

			const element = getByTestId('Heading');

			expect(element).toBeInTheDocument();
		});

		it('should render H1', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H1}>Heading H1</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h1');
		});

		it('should render H2', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H2}>Heading H2</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h2');
		});

		it('should render H3', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H3}>Heading H3</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h3');
		});

		it('should render H4', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H4}>Heading H4</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h4');
		});

		it('should render H5', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H5}>Heading H5</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h5');
		});

		it('should render H6', () => {
			const { getByTestId } = renderWithTheme(<Heading type={HeadingType.H6}>Heading H6</Heading>);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h6');
		});

		it('should render H3 with H2 styles', () => {
			const { getByTestId } = renderWithTheme(
				<Heading as={HeadingType.H3} type={HeadingType.H2}>
					Heading H3 with H2 styles
				</Heading>,
			);

			const element = getByTestId('Heading');

			expect(element).toContainHTML('h3');
		});

		it('should render correctly', () => {
			const tree = renderer.create(<Heading>Heading</Heading>).toJSON();

			expect(tree).toMatchSnapshot();
		});
	});

	describe('Content', () => {
		it('should have valid content', () => {
			const { getByTestId } = renderWithTheme(<Heading>Heading text</Heading>);

			const element = getByTestId('Heading');

			expect(element).toHaveTextContent('Heading text');
		});

		it('should have elements as children', () => {
			const { getByTestId } = renderWithTheme(
				<Heading>
					<span>H</span>eading
				</Heading>,
			);

			const element = getByTestId('Heading');
			const child = element.querySelector('span');

			expect(element).toContainElement(child);
		});
	});
});
