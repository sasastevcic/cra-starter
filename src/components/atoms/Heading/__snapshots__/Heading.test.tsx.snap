// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading Render should render correctly 1`] = `
<h1
  className="sc-bczRLJ bTDTrv"
  data-testid="Heading"
>
  Heading
</h1>
`;
